spark/agent/src/main/scala/za/co/absa/spline/harvester/AttributeDependencySolver.scala

```diff
@@ -0,0 +1,163 @@
+      AttributeDependencies = {
+
+    val attributeDependencies = dataOperations
+      .map(resolveDependencies)
+      .reduce(mergeDependencies)
+
+    val operationDependencies = resolveAttributeOperationDependencies(
+      dataOperations ++ ioOperations,
+      attributeDependencies)
+
+    AttributeDependencies(attributeDependencies, operationDependencies)
+  }
+
+  private def resolveDependencies(op: DataOperation): Map[UUID, Set[UUID]] =
+    op.extra(Name) match {
+      case "Project" => resolveExpressionList(op.params(Transformations), op.schema)
+      case "Aggregate" => resolveExpressionList(op.params(Aggregations), op.schema)
+      case "Generate" => resolveGenerator(op)
+      case _ => Map.empty
+    }
+
+  private def resolveExpressionList(exprList: Any, maybeSchema: Option[Any]): Map[UUID, Set[UUID]] = {
+
+    val schema = maybeSchema.get.asInstanceOf[Seq[UUID]]
+
+    exprList
+      .asInstanceOf[Some[Seq[Expression]]]
+      .get
+      .zip(schema)
+      .map{ case (expr, attrId) => attrId -> toAttrDependencies(expr) }
+      .toMap
+  }
+
+  private def resolveGenerator(op: DataOperation): Map[UUID, Set[UUID]] = {
+
+    val expression = op.params.get("generator").asInstanceOf[Some[Generic]].get
+    val dependencies = toAttrDependencies(expression)
+
+    val keyId = op.params("generatorOutput")
+      .asInstanceOf[Some[Seq[AttrRef]]]
+      .get.head.refId
+
+    Map(keyId -> dependencies)
+  }
+
+  private def toAttrDependencies(expr: Expression): Set[UUID] = expr match {
+    case AttrRef(refId)                        => Set(refId)
+    case Alias(_, child)                       => toAttrDependencies(child)
+    case e: Expression if e.children.nonEmpty  => e.children.map(toAttrDependencies).reduce(_ union _)
+    case _                                     => Set.empty
+  }
+
+  def mergeDependencies(acc: Map[UUID, Set[UUID]], newDependencies:  Map[UUID, Set[UUID]]): Map[UUID, Set[UUID]] =
+    newDependencies.foldLeft(acc) { case (acc, (newKey, newValue)) =>
+
+      // add old dependencies to the new dependencies when they contain one of old keys
+      val addToNewValue = acc.flatMap {
+        case (k, v)  if newValue(k) => v
+        case _                      => Nil
+      }
+
+      val updatedNewValue = newValue.union(addToNewValue.toSet)
+
+      // add new dependencies to all dependencies that contains the new key
+      val updatedAcc = acc.map {
+        case (k, v) if v(newKey) => k -> v.union(updatedNewValue)
+        case (k, v)              => k -> v
+      }
+
+      updatedAcc + (newKey -> updatedNewValue)
+    }
+
+  def schemaOf(op: OperationLike): Seq[UUID] = {
+
+    val maybeSchema = op.schema.asInstanceOf[Option[Seq[UUID]]]
+    maybeSchema.getOrElse(Seq.empty)
+  }
+
+  def schemaMapOf(operations: Seq[OperationLike]): Map[Int, Set[UUID]] = {
+
+    val ioSchemaMap = mutable.Map[Int, Set[UUID]]()
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [It's preferred to use immutable collections instead of mutable ones](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4630741)